public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/setQueryValue.ts

```diff
@@ -109,6 +109,20 @@ export function setMetricNamespace(query: AzureMonitorQuery, metricNamespace: st
     return query;
   }
 
+  let resourceUri = query.azureMonitor?.resourceUri;
+
+  if (resourceUri && metricNamespace?.includes('Microsoft.Storage/storageAccounts')) {
+    const splitUri = resourceUri.split('/');
+    const accountNameIndex = splitUri.findIndex((item) => item === 'storageAccounts') + 1;
+    const baseUri = splitUri.slice(0, accountNameIndex + 1).join('/');
+    if (metricNamespace === 'Microsoft.Storage/storageAccounts') {
+      resourceUri = baseUri;
+    } else {
+      const subNamespace = metricNamespace.split('/')[2];
+      resourceUri = `${baseUri}/${subNamespace}/default`;
+    }
+  }
+
   return {
```

Comment:
I was just reading this https://docs.microsoft.com/en-us/azure/storage/common/storage-account-overview#types-of-storage-accounts and https://docs.microsoft.com/en-us/azure/storage/common/storage-account-create?toc=%2Fazure%2Fstorage%2Fblobs%2Ftoc.json&tabs=azure-portal#create-a-storage-account-1 after reading this part of the code and the way I understand it is that the storage account is the resource, which is identified by the resource uri, and the different storage services (queue, blob, etc.) have different metrics which is the metric namespace. So I'm assuming we don't need to modify the resourceUri here.

I think it should still work without this. I also checked the Azure Portal and copied the link to the dashboard where I'm checking the storage account for a queue capacity metric and the resourceUri it uses doesn't include the sub namespace (storage service) part.

I clicked the copy link and used `decodeURIComponent` to see the resource uri (it's under the property resourceMetadata.id) in the decoded string.

<img width="264" alt="Screen Shot 2022-05-19 at 11 30 09 AM" src="https://user-images.githubusercontent.com/19530599/169374520-b40938f4-86f5-492d-89e5-d595d90b0577.png">

Reply:
I don't think it can, if you take a look at the network requests made when selecting one of the other services besides default you'll see the relative URL will look something like this: 

`"/subscriptions/subId/resourcegroups/rgName/providers/Microsoft.Storage/storageAccounts/acctName/queueServices/default/providers/microsoft.Insights/metrics?timespan=2022-05-19T09:30:00.000Z/2022-05-20T09:30:00.000Z&interval=FULL&metricnames=QueueCapacity&aggregation=average&metricNamespace=microsoft.storage%2Fstorageaccounts%2Fqueueservices&validatedimensions=false&api-version=2019-07-01"`

Similarly, if you take a look at https://docs.microsoft.com/en-us/azure/storage/queues/monitor-queue-storage-reference#resource-logs the `resourceId` field implies that the resource URI needs the service and `default` paths appended. I did test this locally and I couldn't get it working without modifying the resource URI (which is annoying as it also then modifies the display name of the picked resource).

Reply:
Thanks for the explanation. I think it would be worth it to add a comment here to explain why the resourceUri needs to be modified just when it's a storage account.

Reply:
Have done this, thanks @kevinwcyu!